File: src/oop/AutoDestructableV2.c
#include "AutoDestructableV2.r"

void *deleteAllocationAddressNodeFromAllocationTable(
        Legacy_AllocationTable *OBJECT_ALLOCATION_TABLE,
        void *                  allocatedAddress) {
    return OBJECT_ALLOCATION_TABLE->allocationAddressList
            ->deleteNodeThatHasTheGivenData(
                    OBJECT_ALLOCATION_TABLE->allocationAddressList,
                    allocatedAddress);
}

Object *
deleteAllocationAddressIfNeeded(AutoDestructableV2 *autoDestructableV2) {
    if (autoDestructableV2 == NULL) { return NULL; }

    Object *object = autoDestructableV2->allocatedAddress;

    if (object->deleteFromAllocationTableInvocationStatus == WAS_NOT_INVOKED) {
        object->deleteFromAllocationTableInvocationStatus = WAS_INVOKED_ONCE;

        Object *allocatedAddressReturnValue =
                deleteAllocationAddressNodeFromAllocationTable(
                        autoDestructableV2->OBJECT_ALLOCATION_TABLE,
                        autoDestructableV2->allocatedAddress);

        if (allocatedAddressReturnValue == NULL) {

            /*
             * The address was already deleted from
             * `autoDestructableV2->OBJECT_ALLOCATION_TABLE`, and thus had probably already
             * been freed too.
             */
            return NULL;
        }
    }

    return object;
}

void destructAllocatedAddressUnsafe(AutoDestructableV2 *autoDestructableV2) {
    if (autoDestructableV2 == NULL) { return; }

    free(autoDestructableV2->object);
    free(autoDestructableV2);
}

Object *AutoDestructableV2Destructor(AutoDestructableV2 *autoDestructableV2) {
    if (autoDestructableV2 == NULL) { return NULL; }

    if (autoDestructableV2->allocatedAddress->destructorInvocationStatus ==
        WAS_NOT_INVOKED) {
        autoDestructableV2->object->destructorInvocationStatus ==
                WAS_INVOKED_ONCE;

        deleteAllocationAddressIfNeeded(autoDestructableV2);
    }

    if (autoDestructableV2->object->destructorInvocationStatus ==
        WAS_NOT_INVOKED) {
        autoDestructableV2->object->destructorInvocationStatus ==
                WAS_INVOKED_ONCE;

        destructAllocatedAddressUnsafe(autoDestructableV2);
    }
    return NULL;
}

void constructor_AutoDestructableV2_fields(
        AutoDestructableV2 *autoDestructableV2) {
    autoDestructableV2->object = ObjectConstructor();

    static Constructable const constructable = {
            .constructor =
                    (void *(*const)(void) )(&AutoDestructableV2Constructor)};
    autoDestructableV2->object->constructable = &constructable;

    static Destructable const destructable = {
            .destructor =
                    (void *(*const)(void *) )(&AutoDestructableV2Destructor)};
    autoDestructableV2->object->destructable = &destructable;

    autoDestructableV2->object->destructorInvocationStatus = WAS_NOT_INVOKED;
    autoDestructableV2->object->deleteFromAllocationTableInvocationStatus =
            WAS_NOT_INVOKED;
}

void saveObjectToAllocationTable(AutoDestructableV2 *autoDestructableV2) {
    autoDestructableV2->OBJECT_ALLOCATION_TABLE =
            getLegacy_AllocationTableList()
                    ->findLegacy_AllocationTableByClassName(
                            autoDestructableV2->object->CLASS_NAME);
    if (autoDestructableV2->OBJECT_ALLOCATION_TABLE == NULL) {
        autoDestructableV2->OBJECT_ALLOCATION_TABLE =
                Legacy_AllocationTableConstructorWithClassName(
                        (char *) autoDestructableV2->object->CLASS_NAME);

        // Create a legacy_node that its data points to `autoDestructableV2->OBJECT_ALLOCATION_TABLE`.
        Legacy_Node *nodeThatItsDataPointsClassAllocationTable =
                Legacy_NodeConstructorWithDataAndDataSize(
                        autoDestructableV2->OBJECT_ALLOCATION_TABLE,
                        sizeof(Legacy_AllocationTable *));

        // Add this legacy_node to `GLOBAL_ALLOCATION_TABLE_LIST->legacy_allocationTableList`.
        // TODO: best practice to change `add` to `addAsUnique`.
        getLegacy_AllocationTableList()->allocationTableList->add(
                getLegacy_AllocationTableList()->allocationTableList,
                nodeThatItsDataPointsClassAllocationTable);
    }

    // Create a legacy_node that its data points to the "pointer of `autoDestructableV2->allocatedAddress`".
    Legacy_Node *nodeThatItsDataPointsToThePointerOfObj =
            Legacy_NodeConstructorWithDataAndDataSize(
                    autoDestructableV2->allocatedAddress, sizeof(Object *));

    // Add this legacy_node to `autoDestructableV2->OBJECT_ALLOCATION_TABLE->allocationAddressList`.
    // TODO: best practice to change `add` to `addAsUnique`.
    autoDestructableV2->OBJECT_ALLOCATION_TABLE->allocationAddressList->add(
            autoDestructableV2->OBJECT_ALLOCATION_TABLE->allocationAddressList,
            nodeThatItsDataPointsToThePointerOfObj);
}

AutoDestructableV2 *AutoDestructableV2ConstructorWithClassName(
        Object *    objectToSaveItsAddressToAllocationTable,
        const char *className) {
    AutoDestructableV2 *instance = calloc(1, sizeof *instance);
    if (instance == NULL) { /* error handling here */
    }

    constructor_AutoDestructableV2_fields(instance);

    instance->object->CLASS_NAME = className;

    // If `objectToSaveItsAddressToAllocationTable` is `NULL` use `instance`.
    instance->allocatedAddress =
            objectToSaveItsAddressToAllocationTable == NULL
                    ? (Object *) instance
                    : objectToSaveItsAddressToAllocationTable;
    saveObjectToAllocationTable(instance);

    return instance;
}

AutoDestructableV2 *AutoDestructableV2Constructor() {
    return AutoDestructableV2ConstructorWithClassName(NULL,
                                                      "AutoDestructableV2");
}

File: src/oop/AutoDestructableV2.r
#ifndef AUTODESTRUCTABLEV2_H
#define AUTODESTRUCTABLEV2_H

#include "Constructable.r"
#include "Destructable.r"
#include "InvocationStatus.r"
#include "Legacy_AllocationTable.r"
#include "Legacy_AllocationTableList.r"
#include "Object.r"
#include "ObjectContainer.r"
#include <stdio.h>
#include <stdlib.h>

// Forward declaration of incomplete type
typedef struct autoDestructableV2 AutoDestructableV2;

struct autoDestructableV2 {

    /// `AutoDestructableV2` implements `Object`.
    Object *object;

    /// Singleton for the whole autoDestructableV2. Sensitive data. DO NOT TOUCH!
    Legacy_AllocationTable *OBJECT_ALLOCATION_TABLE;

    /// Sensitive data. DO NOT TOUCH!
    Object *allocatedAddress;
};

AutoDestructableV2 *AutoDestructableV2ConstructorWithClassName(
        Object *objectToSaveItsAddressToAllocationTable, const char *className);

/**
 * @brief Creates a `new` "heap-allocated" instance of `AutoDestructableV2` and
 *        initializes it.
 *
 * @attention Remember to `free` the returned value of this function.
 *
 * For example:
 * @code
 *
 * // Construct.
 * AutoDestructableV2 *autoDestructableV2 = AutoDestructableV2Constructor();
 *
 * // Do something...
 * (*autoDestructableV2->print)(autoDestructableV2);
 * (*autoDestructableV2->addOneToX)(autoDestructableV2);
 * (*autoDestructableV2->print)(autoDestructableV2);
 *
 * // Remember to `free`.
 * free(autoDestructableV2);
 * @endcode
 *
 * @return A pointer to a `new` "heap-allocated" instance of `AutoDestructableV2`.
 */
AutoDestructableV2 *AutoDestructableV2Constructor();

Object *AutoDestructableV2Destructor(AutoDestructableV2 *autoDestructableV2);

#endif //AUTODESTRUCTABLEV2_H
